### PR TITLE
ll-schedule: fix scheduling start on slave cores

### PR DESCRIPTION
**Pull request by lyakh**
This fixes a recently reported and discussed xrun on slave cores.

### CODE REVIEW

File: src/schedule/ll_schedule.c [hunk 1]
```diff
@@ -200,11 +201,14 @@ static int schedule_ll_domain_set(struct ll_schedule_data *sch,
...
 
+	registered = sch->domain->registered[core];
 	if (atomic_add(&sch->num_tasks, 1) == 1)
 		sch->domain->registered[core] = true;
 
-	if (atomic_add(&sch->domain->total_num_tasks, 1) == 1) {
+	if (atomic_add(&sch->domain->total_num_tasks, 1) == 1)
 		domain_set(sch->domain, platform_timer_get(timer_get()));
+
+	if (!registered) {
```
Comment (lgirdwood) [inline src/schedule/ll_schedule.c]: Looks fine to me, I would just comment the logic here so that it's obvious why we are doing this

**Comment by tlauda** [comment]
@jdrzaszcz Please run multicore stress tests for this patch. Let's make sure it doesn't break logic somewhere else in the scheduler.

**Comment by jdrzaszcz** [comment]
I run the multicore stress tests on SSP loopback (same SSP) with pipelines run on two different cores. 
Playback pipeline on core 0.
Capture pipeline on core 1.
I run this test with playback_iterations= 100. 
Tests fail in 35 cases on 48.
I got stream stall when sending trigger_start on playback pipeline in different playback_iteration depends on test case (format).
Tests were made for master commit f1fa3be9804f0a51e4b80b1f1c28a00973844e54 + this PR

When I run the same tests only for a master branch, commit: f1fa3be9804f0a51e4b80b1f1c28a00973844e54, I didn't see this problem.

**Comment by lyakh** [comment]
> I run the multicore stress tests on SSP loopback (same SSP) with pipelines run on two different cores.
> Playback pipeline on core 0.
> Capture pipeline on core 1.
> I run this test with playback_iterations= 100.
> Tests fail in 35 cases on 48.
> I got stream stall when sending trigger_start on playback pipeline in different playback_iteration depends on test case (format).
> Tests were made for master commit [f1fa3be](https://github.com/thesofproject/sof/commit/f1fa3be9804f0a51e4b80b1f1c28a00973844e54) + this PR
> 
> When I run the same tests only for a master branch, commit: [f1fa3be](https://github.com/thesofproject/sof/commit/f1fa3be9804f0a51e4b80b1f1c28a00973844e54), I didn't see this problem.

@jdrzaszcz is this under Linux or with the slim driver? If under Linux - how can I deproduce it? If with the slim driver - could you send me DSP traces from a successful and a failing run, please?

**Comment by jdrzaszcz** [comment]
Sorry for late replay, but I wasn't at work at Friday.
Tests were on slim driver. I attached logs.
[dma_trace_1_fail.log](https://github.com/thesofproject/sof/files/4367505/dma_trace_1_fail.log)
[dma_trace_1_pass.log](https://github.com/thesofproject/sof/files/4367506/dma_trace_1_pass.log)

**Comment by lyakh** [comment]
> Sorry for late replay, but I wasn't at work at Friday.
> Tests were on slim driver. I attached logs.
> [dma_trace_1_fail.log](https://github.com/thesofproject/sof/files/4367505/dma_trace_1_fail.log)
> [dma_trace_1_pass.log](https://github.com/thesofproject/sof/files/4367506/dma_trace_1_pass.log)

@jdrzaszcz thanks for the logs. It seems the configuration between the two test cases is also different though:
1. buffer size pass 256B, fail 128B
2. frame format pass 0 fail 2
3. sample rate pass 16000 fail 8000
4. channels pass 4 fail 2
5. width pass 2 fail 4

why?

**Comment by jdrzaszcz** [comment]
The tests use different audio formats.

**Comment by jdrzaszcz** [comment]
@lyakh answer above

**Comment by lyakh** [comment]
> The tests use different audio formats.

@jdrzaszcz sorry, but then I cannot compare them. Could you run the *same* test with and without my patch? Do you then still see *no* failures without my patch? Could you then send me logs: (1) success without my patch, (2) success with my patch, (3) failure with my patch.

**Comment by jdrzaszcz** [comment]
@lyakh All tests must be in the same audio format, right?

**Comment by lyakh** [comment]
> @lyakh All tests must be in the same audio format, right?

@jdrzaszcz yes, please!

**Comment by jdrzaszcz** [comment]
@lyakh  dma_trace files attached
[dma_trace_1_master_pass.log](https://github.com/thesofproject/sof/files/4368583/dma_trace_1_master_pass.log)
[dma_trace_1_sink_master_pass.log](https://github.com/thesofproject/sof/files/4368584/dma_trace_1_sink_master_pass.log)
[dma_trace_1_sink_path_pass.log](https://github.com/thesofproject/sof/files/4368586/dma_trace_1_sink_path_pass.log)
[dma_trace_1_path_pass.log](https://github.com/thesofproject/sof/files/4368587/dma_trace_1_path_pass.log)
[dma_trace_1_sink_path_fail.log](https://github.com/thesofproject/sof/files/4368589/dma_trace_1_sink_path_fail.log)
[dma_trace_1_path_fail.log](https://github.com/thesofproject/sof/files/4368591/dma_trace_1_path_fail.log)

**Comment by lyakh** [comment]
> @lyakh dma_trace files attached
> [dma_trace_1_master_pass.log](https://github.com/thesofproject/sof/files/4368583/dma_trace_1_master_pass.log)
> [dma_trace_1_sink_master_pass.log](https://github.com/thesofproject/sof/files/4368584/dma_trace_1_sink_master_pass.log)
> [dma_trace_1_sink_path_pass.log](https://github.com/thesofproject/sof/files/4368586/dma_trace_1_sink_path_pass.log)
> [dma_trace_1_path_pass.log](https://github.com/thesofproject/sof/files/4368587/dma_trace_1_path_pass.log)
> [dma_trace_1_sink_path_fail.log](https://github.com/thesofproject/sof/files/4368589/dma_trace_1_sink_path_fail.log)
> [dma_trace_1_path_fail.log](https://github.com/thesofproject/sof/files/4368591/dma_trace_1_path_fail.log)

@jdrzaszcz thanks for the logs! Could you please explain a bit what names mean? "master" means the current tip of the master branch, right? Without "master" means with my PR applied? What do "path" and "sink" mean?

Also, could you please confirm, that now with the same configuration the master branch still doesn't fail?

**Comment by jdrzaszcz** [comment]
@lyakh Yes, master is master branch, but path means master + your PR.
In sink file, as far as I know,  are all data read from DMA buffer in case  all data were not saved to the dma_trace file.

Yes, the test passes for today's master branch.

**Comment by lyakh** [comment]
@jdrzaszcz could you please re-test? If this still fails, I'll have to try and run the slim-driver myself

**Comment by jdrzaszcz** [comment]
@lyakh I re-tested this PR with master branch and still fail.

**Comment by lyakh** [comment]
> @lyakh I re-tested this PR with master branch and still fail.

@jdrzaszcz ok, thanks... If you have any hints how to execute this test - please let me know.

**Comment by lyakh** [comment]
> > @lyakh I re-tested this PR with master branch and still fail.
> 
> @jdrzaszcz ok, thanks... If you have any hints how to execute this test - please let me know.

@jdrzaszcz is it the fw_17_multi_core test that failed? Unmodified or did you have to change anything for specific DAIs and formats?

**Comment by paulstelian97** [comment]
@jdrzaszcz Editing @lyakh's comment to give him a response isn't very effective as edits don't send notifications even to the original author on Github.

**Comment by jdrzaszcz** [comment]
@paulstelian97 Thank you for the tip and editing my comment.

**Comment by lyakh** [comment]
> > > @lyakh I re-tested this PR with master branch and still fail.
> > 
> > 
> > @jdrzaszcz ok, thanks... If you have any hints how to execute this test - please let me know.
> 
> @jdrzaszcz is it the fw_17_multi_core test that failed? Unmodified or did you have to change anything for specific DAIs and formats?

I finally managed to run the fw_17_multi_core/test_01_multi_core_multi_kernel_ssp.py test on a laptop but I couldn't reproduce the problem. I know that @jdrzaszcz is now on holiday for a week, maybe someone else can help me? Was the test just run unmodified or have any adjustments been done to set specific formats? I used the `--force-ssp-internal` flag to make running on a laptop possible, could that have changed the result? I'll try on an Up^2 next week.

**Comment by lyakh** [comment]
> > > > @lyakh I re-tested this PR with master branch and still fail.
> > > 
> > > 
> > > @jdrzaszcz ok, thanks... If you have any hints how to execute this test - please let me know.
> > 
> > 
> > @jdrzaszcz is it the fw_17_multi_core test that failed? Unmodified or did you have to change anything for specific DAIs and formats?
> 
> I finally managed to run the fw_17_multi_core/test_01_multi_core_multi_kernel_ssp.py test on a laptop but I couldn't reproduce the problem. I know that @jdrzaszcz is now on holiday for a week, maybe someone else can help me? Was the test just run unmodified or have any adjustments been done to set specific formats? I used the `--force-ssp-internal` flag to make running on a laptop possible, could that have changed the result? I'll try on an Up^2 next week.

Still cannot reproduce the problem - on Up^2 too. Could anyone help me find out exactly what and how has been tested?

**Comment by lgirdwood** [comment]
@lbetlej can anyone help or should we wait for @jdrzaszcz to return ?

**Comment by lyakh** [comment]
> @lyakh I re-tested this PR with master branch and still fail.

@jdrzaszcz this version passes my testing - both under Linux and with pytest. Would be cool if you could give it a try.

**Comment by jdrzaszcz** [comment]
@lyakh ok, i will run the tests and let you know.

**Comment by lgirdwood** [comment]
Probably too late for v1.5, so lets take in v1.6

**Comment by jdrzaszcz** [comment]
@lyakh Hi, I have run the multi_core_multi_kernel_ssp test with your pull request and the test have passed for 100 playback iteration in all 48 formats.

**Comment by lyakh** [comment]
> @lyakh Hi, I have run the multi_core_multi_kernel_ssp test with your pull request and the test have passed for 100 playback iteration in all 48 formats.

@jdrzaszcz thanks!
@tlauda @lgirdwood not sure about CI failures. The device test seems to be unrelated, as for the "codecheck" - I couldn't actually identify what it was complaining about?

**Comment by lgirdwood** [comment]
@lyakh looks like known CI issues on jenkins

**Pull request merged**